File: cli/python/ime_control/komorebi.py
from __future__ import annotations

import shutil
import subprocess
from dataclasses import dataclass


DEFAULT_WINDOW_KINDS = ("single", "stack", "monocle", "floating")


@dataclass(frozen=True)
class RgbColor:
    red: int
    green: int
    blue: int

    @classmethod
    def parse(cls, text: str) -> "RgbColor":
        parts = [part.strip() for part in text.split(",")]
        if len(parts) != 3:
            raise ValueError(f"RGB color must have 3 comma-separated parts, got '{text}'")
        values = [int(part) for part in parts]
        if any(value < 0 or value > 255 for value in values):
            raise ValueError(f"RGB color values must be in 0..255, got '{text}'")
        return cls(*values)

    def as_args(self) -> list[str]:
        return [str(self.red), str(self.green), str(self.blue)]


def resolve_komorebic(explicit_path: str | None = None) -> str:
    if explicit_path:
        return explicit_path
    path = shutil.which("komorebic.exe") or shutil.which("komorebic")
    if path is None:
        raise FileNotFoundError("komorebic executable was not found on PATH")
    return path


def apply_border_colours(
    *,
    komorebic_path: str,
    color: RgbColor,
    window_kinds: tuple[str, ...] = DEFAULT_WINDOW_KINDS,
) -> None:
    for window_kind in window_kinds:
        subprocess.run(
            [komorebic_path, "border-colour", "--window-kind", window_kind, *color.as_args()],
            check=True,
            capture_output=True,
            text=True,
        )

File: src/komorebi.rs
use std::env;
use std::ffi::OsString;
use std::path::{Path, PathBuf};
use std::process::Command;

use anyhow::{Context, Result};

pub const DEFAULT_WINDOW_KINDS: &[&str] = &["single", "stack", "monocle", "floating"];

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct RgbColor {
    pub red: u8,
    pub green: u8,
    pub blue: u8,
}

impl RgbColor {
    pub fn parse(text: &str) -> Result<Self> {
        let parts: Vec<&str> = text.split(',').map(str::trim).collect();
        if parts.len() != 3 {
            anyhow::bail!("RGB color must have 3 comma-separated parts, got '{text}'");
        }
        let red = parts[0]
            .parse::<u8>()
            .with_context(|| format!("invalid red component in '{text}'"))?;
        let green = parts[1]
            .parse::<u8>()
            .with_context(|| format!("invalid green component in '{text}'"))?;
        let blue = parts[2]
            .parse::<u8>()
            .with_context(|| format!("invalid blue component in '{text}'"))?;
        Ok(Self { red, green, blue })
    }

    pub fn as_args(&self) -> [String; 3] {
        [
            self.red.to_string(),
            self.green.to_string(),
            self.blue.to_string(),
        ]
    }
}

pub fn resolve_komorebic(explicit_path: Option<&Path>) -> Result<PathBuf> {
    if let Some(path) = explicit_path {
        return Ok(path.to_path_buf());
    }
    find_on_path("komorebic.exe")
        .or_else(|| find_on_path("komorebic"))
        .ok_or_else(|| anyhow::anyhow!("komorebic executable was not found on PATH"))
}

fn find_on_path(name: &str) -> Option<PathBuf> {
    let path_var = env::var_os("PATH")?;
    env::split_paths(&path_var)
        .map(|dir| dir.join(name))
        .find(|candidate| candidate.is_file())
}

pub fn border_colour_command_args(window_kind: &str, color: RgbColor) -> Vec<OsString> {
    let rgb = color.as_args();
    vec![
        OsString::from("border-colour"),
        OsString::from("--window-kind"),
        OsString::from(window_kind),
        OsString::from(&rgb[0]),
        OsString::from(&rgb[1]),
        OsString::from(&rgb[2]),
    ]
}

pub fn apply_border_colours(
    komorebic_path: &Path,
    color: RgbColor,
    window_kinds: &[String],
) -> Result<()> {
    apply_border_colours_with(komorebic_path, color, window_kinds, |program, args| {
        let status = Command::new(program)
            .args(args)
            .status()
            .with_context(|| format!("failed to execute {}", program.display()))?;
        if !status.success() {
            anyhow::bail!("komorebic exited with status {status}");
        }
        Ok(())
    })
}

pub fn apply_border_colours_with<F>(
    komorebic_path: &Path,
    color: RgbColor,
    window_kinds: &[String],
    mut runner: F,
) -> Result<()>
where
    F: FnMut(&Path, &[OsString]) -> Result<()>,
{
    for window_kind in window_kinds {
        let args = border_colour_command_args(window_kind, color);
        runner(komorebic_path, &args)?;
    }
    Ok(())
}
